src/mscorlib/shared/System/Memory.cs

```diff
@@ -290,9 +292,12 @@ public Span<T> Span
                     return new Span<T>(ref Unsafe.As<char, T>(ref s.GetRawStringData()), s.Length).Slice(_index, _length);
 #endif // FEATURE_PORTABLE_SPAN
                 }
-                else if (_object != null)
+                else if (obj != null)
                 {
-                    return new Span<T>((T[])_object, _index, _length & RemoveFlagsBitMask);
+                    // obj is gauranteed to be a T[] here,
+                    // which is why we can use Unsafe.As as a performance optimization instead of an explicit cast
+                    Debug.Assert(obj is T[]);
+                    return new Span<T>(Unsafe.As<T[]>(obj), _index, _length & RemoveFlagsBitMask);
```

Comment:
I know we walked through this in person earlier today, but the more I think about this I fear our original analysis was incorrect. :(

There could possibly be a scenario where `_index >= 0` and `obj` is a `MemoryManager<T>` due to a tear, which makes this unsafe cast dangerous. So we might need to keep the proper cast after all. The reason `ValueTask<T>` was able to work around it is that the code in those methods only ever checks a single local, whereas here we're checking two values (_obj_ and _index_) which could get out of sync.

Reply:
Was the cost of an `is Type var` cast compared to a regular `(Type)` cast?

If they are comparable, then it may be worth dropping the flags on `_index` doing the first cast as an `if (obj is MemoryManager<T> manager)` cast then the last cast to `T[]` can be `Unsafe.As` as `string` has also been checked with `is` so the last type check is implicit and there is no tearing of type? (Though may still be of index vs obj)

Reply:
Right, this does not look safe to me.

Reply:
i.e. if `(Type)` cast is used then using the flag saves one type test

* If `index < 0` cast to `(MemoryManager<T>)obj` (1 cast test)
* else cast to `(T[])obj` (1 cast test)

Whereas, the flagging of the index could be dropped and the cast tests would be the same?

* If `obj is MemoryManager<T> manager` (1 cast test)
* else `Unsafe.As<T[]>(obj)` (1 cast test, previous)

Reply:
@benaadams' approach is effectively what I did in ValueTask.  You have N possible things that can be stored in a field.  You copy that field to a local, and then type-test that local for N-1 of them; you can then use Unsafe.As for the Nth, as the only way it could be something else at that point is if unsafe/marshaling/private reflection code were used, at which point type safety could already be violated.

> The reason ValueTask<T> was able to work around it is that the code in those methods only ever checks a single local

Right.  Any decisions to avoid a cast and use Unsafe.As needs to be based on a single non-tearable read.